src/main/java/frc/robot/subsystems/SwerveModule.java

```diff
@@ -0,0 +1,147 @@
+
+    private final PIDController turningPIDController;
+
+    private final WPI_CANCoder absoluteEncoder;
+
+    private final String ID;
+
+    public SwerveModule(MotorConfig driveMotorConfig, MotorConfig turningMotorConfig, EncoderConfig absoluteEncoderConfig, String ID) {
+
+        this.ID = ID;
+
+        absoluteEncoder = Robot.isSimulation() ? null : AbsoluteEncoder.constructEncoder(absoluteEncoderConfig);
+
+        driveMotor = MotorController.constructMotor(driveMotorConfig);
+        turningMotor = MotorController.constructMotor(turningMotorConfig);
+
+        driveEncoder = this.driveMotor.getEncoder();
+        turningEncoder = this.turningMotor.getEncoder();
+
+        driveEncoder.setPositionConversionFactor(SwerveModuleConstants.kDriveEncoderRotFactor);
+        driveEncoder.setVelocityConversionFactor(SwerveModuleConstants.kDriveEncoderRPMFactor);
+        turningEncoder.setPositionConversionFactor(SwerveModuleConstants.kTurningEncoderRotFactor);
+        turningEncoder.setVelocityConversionFactor(SwerveModuleConstants.kTurningEncoderRPMFactor);
+
+        simDriveMotor = new FlywheelSim(
+            LinearSystemId.identifyVelocitySystem(2, 1.24), //TODO: Update with real SysID
+            DCMotor.getNEO(1),
+            SwerveModuleConstants.kDriveMotorGearRatio
+        );
+        simTurningMotor = new FlywheelSim(
+            LinearSystemId.identifyVelocitySystem(0.16, 0.0348), //TODO: Update with real SysID
+            DCMotor.getNEO(1),
+            SwerveModuleConstants.kTurningMotorGearRatio
+        );
+
+        simDriveEncoder = new RelativeEncoderSim(driveMotor);
+        simTurningEncoder = new RelativeEncoderSim(turningMotor);
+        
+        turningPIDController = new PIDController(SwerveModuleConstants.kPTurning, 0, 0);
+        turningPIDController.enableContinuousInput(-Math.PI, Math.PI);
+
+        resetEncoders();
+    }
+
+    public double getDrivePosition() {
+        return driveEncoder.getPosition();
+    }
+
+    public double getTurningPosition() {
+        return turningEncoder.getPosition();
+    }
+
+    public double getDriveVelocity() {
+        return Robot.isSimulation() ? simDriveEncoder.getVelocity() : driveEncoder.getVelocity();
+    }
+    
+
+    public double getTurningVelocity() {
+        return Robot.isSimulation() ? simTurningEncoder.getVelocity() : turningEncoder.getVelocity();
+    }
+
+    public void resetEncoders() {
+        driveEncoder.setPosition(0);
+        turningEncoder.setPosition(Robot.isSimulation() ? 0 : Units.degreesToRotations(absoluteEncoder.getAbsolutePosition()));
+    }
+
+    public SwerveModuleState getState() {
+        return new SwerveModuleState(getDriveVelocity(), new Rotation2d(getTurningPosition()));
+    }
+
+    public void setDesiredState(SwerveModuleState state) {
+        if (Math.abs(state.speedMetersPerSecond) < 0.001) { //Prevent wheels to returning to heading of 0 when controls released
```

Comment:
This is a problem. You are checking the state's wheel speed and then stopping both the turning motor and the driving motor if the wheel speed is 0. This probably leads to a lot of unintended affects.

convertToModuleStates constructs the SwerveModuleState[] array that eventually sends a single state to each module.
convertToModuleState passes calls [ChassisSpeeds.fromFieldRelativeSpeeds(...)](https://github.wpilib.org/allwpilib/docs/release/java/edu/wpi/first/math/kinematics/ChassisSpeeds.html#fromFieldRelativeSpeeds(double,double,double,edu.wpi.first.math.geometry.Rotation2d))

This is the desired speed we want to rotate at. So if the joystick value is showing a zero that means it will try to make the robot rotate 0 omegaRadiansPerSecond for the next status.

Reply:
A lot of the testing we were doing was with 0 x,y position change and just angular change. So if our wheel speed was 0 we would stop all motors and return.

Reply:
> checking the state's wheel speed and then stopping both the turning motor and the driving motor if the wheel speed is 0

Shouldn't the robot be stopping when it is commanded to stop?

> A lot of the testing we were doing was with 0 x,y position change and just angular change

I was testing almost exclusively with translational inputs. (Although rotational input would still have a high drive motor speed)

>if the joystick value is showing a zero that means it will try to make the robot rotate 0 omegaRadiansPerSecond for the next status.

Is this a problem? If the joystick commands zero on the rotation axis, why would we want the robot to continue to rotate?


I was testing today, and all modules were performing much better without any oscillation after the hardware was fixed

Reply:
> Is this a problem? If the joystick commands zero on the rotation axis, why would we want the robot to continue to rotate?
No that's fine I was outlining my reasoning. 

>>  checking the state's wheel speed and then stopping both the turning motor and the driving motor if the wheel speed is 0

> Shouldn't the robot be stopping when it is commanded to stop?

The problem is that you are stopping both the rotation motors and the drive motors, but you are only checking if the drive motor speed is about 0.0 to stop both.

If a state was provided that had a wheel speed of 0.0, and a rotation component right now it wouldn't work.  So if you wanted to set the modules to a rotation, but not move yet, its not possible with this implementation.

Reply:
> If a state was provided that had a wheel speed of 0.0, and a rotation component right now it wouldn't work

This isn't possible. Modules will never rotate without also moving.

Reply:
> This isn't possible. Modules will never turn without also moving.

That is incorrect.
See https://docs.wpilib.org/en/stable/docs/software/kinematics-and-odometry/swerve-drive-kinematics.html#using-custom-centers-of-rotation

Reply:
> the comment didn't match what was happening in the code

I don't think that's true. kDriveKinematics.toSwerveModuleStates will default the heading of each module to zero whenever zero is provided for chassis speeds, as seen [here](https://github.com/wpilibsuite/allwpilib/blob/b8d019cdb47ce4b091b92f91637d512b7d800f26/wpimath/src/main/java/edu/wpi/first/math/kinematics/SwerveDriveKinematics.java#L123)
This is verified by simulation, all modules snap back to a heading of zero when I stop providing input.
I don't think we want our drive to be doing this, and so that deadzone prevents this.

> Also consider the use case in auto. You align to something like the hub for scoring, while you are scoring you could align your wheels to the next step to be ready to move.

The time swerve takes to turn modules is short enough that this isn't necessary, but sure, that is a use case
I think that if we need this as an ability of swerve, it should be a separate method, so that we are not sacrificing any controllability in teleop.
That separate method would be cumbersome to implement, as trajectories are built from many Pose2d points, which contain x, y and rotation states for the drivebase as a whole. So, we would need to get these three values by sampling the initial point and the point ~0.1 seconds into the next trajectory, find the distance between these values and the difference in rotation, and then divide these distances by the time we used to get chassis speeds. Then, we need to convert these speeds to module states, and only use the heading of these states.
It could be done though, if we want the method.

> That might not always be true.
See https://docs.wpilib.org/en/stable/docs/software/kinematics-and-odometry/swerve-drive-kinematics.html#using-custom-centers-of-rotation

Custom centers of rotation still send substantial velocities to all modules
Verified with simulation, setting the center of rotation exactly in the middle of a module

Reply:
Alright this was a deep dig to figure out why it'd reset the angle if provided a value of 0 for rotation input.

Fine to keep it for now, in 2023 when we upgrade it won't be needed as they fixed a bug /made a change in WPILib.
https://github.com/wpilibsuite/allwpilib/pull/4208

It was very confusing reading all of the documentation and seeing no mention of swerve state angle getting reset in toSwerveModuleStates.

Good job on doing the simulation to find out about center of rotations.